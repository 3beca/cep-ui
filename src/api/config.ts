export const BASE_URL = process.env.REACT_APP_BASE_URL || 'https://cep.tribeca.ovh';
export const EVENT_TYPE_URL = '/admin/event-types';