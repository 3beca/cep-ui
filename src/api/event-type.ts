import {
    fetchData,
    RequestInfo,
    ResponseData,
    ErrorAPI,
    ResponseEmptyData
} from './fetch-data';
import {
    BASE_URL,
    EVENT_TYPE_URL
} from './config';
export * from './fetch-data';

export type EventType = {
    name: string;
    id: string;
    url: string;
    createdAt: string;
    updatedAt: string;
};
export type EventTypeList = {
    results: EventType[];
    next?: string | undefined;
    prev?: string | undefined;
};
export const getEventTypeList = (config: RequestInfo) => async (page: number = 1, size: number = 10): Promise<ResponseData<EventTypeList>|ErrorAPI> => {
    const url = `${config.url}${EVENT_TYPE_URL}/?page=${page}&pageSize=${size}`;
    return await fetchData<undefined, EventTypeList>({...config, url, method: 'GET'});
};
export const deleteEventType = (config: RequestInfo) => async (eventTypeIds: string|string[]): Promise<(ResponseEmptyData|ErrorAPI)[]> => {
    const eventTypesArray = (Array.isArray(eventTypeIds)) ? eventTypeIds : [eventTypeIds];
    if(!eventTypeIds || eventTypesArray.length < 1) {
        return [{
            status: 500,
            error: 'Missing EventType id',
            message: 'eventTypeIds is an invalid EventTypeId value or array'
        }];
    }
    const requests = [];
    for (const eventTypeId of eventTypesArray) {
        const url = `${config.url}${EVENT_TYPE_URL}/${eventTypeId}`;
        requests.push(fetchData<undefined, undefined>({...config, url, method: 'DELETE'}));
    }
    return Promise.all(requests);
};
export type EventTypeApi = {
    getEventTypeList(page?: number, size?: number): Promise<ResponseData<EventTypeList>|ErrorAPI>;
    deleteEventType(eventTypeIds: string|string[]): Promise<(ResponseEmptyData|ErrorAPI)[]>;
};
export const EventTypeApi = (): EventTypeApi => {
    const config: RequestInfo = {
        url: BASE_URL,
        method: 'GET',
        headers: {}
    };
    return {
        getEventTypeList: getEventTypeList(config),
        deleteEventType: deleteEventType(config)
    };
};
