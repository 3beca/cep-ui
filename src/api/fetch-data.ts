export type APIHeaders = {[key:string]: string;}; // TODO: restrict to valid headers for CEP
export type RequestInfo = {
    url: string,
    method: 'GET'|'POST'|'PUT'|'DELETE';
    headers: APIHeaders;
};
export type Body = object | undefined;
export type RequestData<T extends Body> = {
    body?: T;
} & RequestInfo;
export type RequestGetData = RequestData<undefined>;
export type ErrorAPI = {
    status: number;
    error: string;
    message: string;
};
export type ResponseData<T> = {
    status: number;
    data: T;
};
export type ResponseEmptyData = ResponseData<undefined>;
export const fetchData = async <B extends Body, R extends Body>(config: RequestData<B>): Promise<ResponseData<R> | ErrorAPI> => {
    const {url, method, body, ...rest} = config;
    const options: RequestInit = {
        method,
        body: (method !== 'GET' && body) ? JSON.stringify(body) : undefined,
        ...rest
    };
    
    try {
        const response = await fetch(url, options);
        const status = response.status;
        let data = undefined;
        try {
            data = await response.json();
        }
        catch (error) {

        }
        if(response.ok) {
            return {
                status,
                data
            };
        }
        else {
            return {
                status: data.statusCode || status,
                error: data.error || `Error from ${url}`,
                message: data.message
            };
        }
    }
    catch(error) {
        return {
            status: 500,
            error: 'Invalid request ' + url,
            message: error.message
        };
    }
};

export const isErrorApi = <T>(error: ResponseData<T>|ErrorAPI): error is ErrorAPI => {
    return (error as ErrorAPI).error !== undefined;
}
