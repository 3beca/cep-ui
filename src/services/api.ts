import {
    fetchApi,
    APIRequestInfo,
    APIResponseData,
    APIError,
    isAPIError
} from '../utils/fetch-api';

export type ServiceList<T> = {
    results: T[];
    next?: string | undefined;
    prev?: string | undefined;
};
export type ServiceError = {
    statusCode: number,
    error: string,
    message: string
};
export type Entity = {
    id: string;
};
export type WithDate = {
    createdAt: string;
    updatedAt: string;
};
export type ServiceDeleted = {
    state: 'DELETED'|'REJECTED';
    error?: ServiceError;
} & Entity;
export type EventType = {
    name: string;
    url: string;
} & Entity & WithDate;
export type EventTypeList = ServiceList<EventType>;
export type EventTypeError = ServiceError;
export type EventTypeDeleted = ServiceDeleted;

export type Target = {
    name: string;
    url: string;
} & Entity & WithDate;
export type TargetList = ServiceList<Target>;
export type TargetError = ServiceError;
export type TargetDeleted = ServiceDeleted;

export type RuleTypes = 'sliding'|'hopping'|'tumbling'|'none';
export type Rule = {
    name: string;
    type: RuleTypes;
    targetId: string;
    eventTypeId: string;
    skipOnConsecutivesMatches?: boolean;
    filters: any;
} & Entity & WithDate;
export type RuleList = ServiceList<Rule>;
export type RuleError = ServiceError;
export type RuleDeleted = ServiceDeleted;

export type EventLog = {
    eventTypeId: string;
    eventTypeName: string;
    payload: {[key: string]: any;};
    requestId: string;
    createdAt: string;
    rules: {
        id: string;
        name: string;
        targetId: string;
    }[];
    targets: {
        id: string;
        name: string;
        response: {
            statusCode: number;
            body: {[key: string]: any;};
        };
    }[];
} & Entity;

export const getListRequest = (baseURL: string, config: APIRequestInfo) => async <T>(path: string, page: number = 1, size: number = 10, filter: string = ''): Promise<APIResponseData<ServiceList<T>>|APIError<ServiceError>> => {
    const url = `${baseURL}${path}/?page=${page}&pageSize=${size}${filter ? `&search=${filter}` : ''}`;
    return fetchApi<undefined, ServiceList<T>, ServiceError>(url, {...config, method: 'GET'});
};
export const deleteRequest = (baseURL: string, config: APIRequestInfo) => async (path: string, eventTypeIds: string|string[]): Promise<(APIResponseData<ServiceDeleted[]>|APIError<ServiceError>)> => {
    const eventTypesArray = ((Array.isArray(eventTypeIds)) ? eventTypeIds : [eventTypeIds]).filter(id => !!id);

    if(!eventTypeIds || eventTypesArray.length < 1) {
        return {
            errorCode: 500,
            errorMessage: '',
            error: {
                statusCode: 500,
                error: 'Missing id',
                message: 'id is an invalid id value or array'
            }
        };
    }
    const requests = [];
    for (const eventTypeId of eventTypesArray) {
        const url = `${baseURL}${path}/${eventTypeId}`;
        requests.push(fetchApi<undefined, undefined, ServiceError>(url, {...config, method: 'DELETE'}));
    }
    const responses = await Promise.all(requests);
    const data: ServiceDeleted[] = responses.map((response, idx) => {
        const deletedResponse: ServiceDeleted = {
            id: eventTypesArray[idx],
            state: 'DELETED'
        };
        if(isAPIError(response)) {
            deletedResponse.state = 'REJECTED';
            deletedResponse.error = response.error;
        }
        return deletedResponse;
    });
    return {
        status: 200,
        data
    };
};
export type Api = {
    getListRequest<T>(path: string, page?: number, size?: number, filter?: string): Promise<APIResponseData<ServiceList<T>>|APIError<ServiceError>>;
    deleteRequest(path: string, ids: string|string[]): Promise<(APIResponseData<ServiceDeleted[]>|APIError<ServiceError>)>;
};
export const buildApiService = (server: string): Api => {
    const config: APIRequestInfo = {
        method: 'GET',
        headers: {}
    };
    return {
        getListRequest: getListRequest(server, config),
        deleteRequest: deleteRequest(server, config)
    };
};
