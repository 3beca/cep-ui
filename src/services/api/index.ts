import {
    fetchApi,
    APIRequestInfo,
    APIResponseData,
    APIError,
    isAPIError,
    APIBody
} from '../../utils/fetch-api';
import {
    ServiceList,
    ServiceError,
    ServiceDeleted,
} from './models';
export * from './models';

export const getListRequest = (baseURL: string, config: APIRequestInfo) => async <T>(path: string, page: number = 1, size: number = 10, filter: string = ''): Promise<APIResponseData<ServiceList<T>>|APIError<ServiceError>> => {
    const url = `${baseURL}${path}/?page=${page}&pageSize=${size}${filter ? `&search=${filter}` : ''}`;
    return fetchApi<undefined, ServiceList<T>, ServiceError>(url, {...config, method: 'GET'});
};
export const deleteRequest = (baseURL: string, config: APIRequestInfo) => async (path: string, entityIds: string|string[]): Promise<(APIResponseData<ServiceDeleted[]>|APIError<ServiceError>)> => {
    const entityArray = ((Array.isArray(entityIds)) ? entityIds : [entityIds]).filter(id => !!id);

    if(!entityIds || entityArray.length < 1) {
        return {
            errorCode: 500,
            errorMessage: '',
            error: {
                statusCode: 500,
                error: 'Missing id',
                message: 'id is an invalid id value or array'
            }
        };
    }
    const requests = [];
    for (const entityId of entityArray) {
        const url = `${baseURL}${path}/${entityId}`;
        requests.push(fetchApi<undefined, undefined, ServiceError>(url, {...config, method: 'DELETE'}));
    }
    const responses = await Promise.all(requests);
    const data: ServiceDeleted[] = responses.map((response, idx) => {
        const deletedResponse: ServiceDeleted = {
            id: entityArray[idx],
            state: 'DELETED'
        };
        if(isAPIError(response)) {
            deletedResponse.state = 'REJECTED';
            deletedResponse.error = response.error;
        }
        return deletedResponse;
    });
    return {
        status: 200,
        data
    };
};
export const createRequest = (baseURL: string, config: APIRequestInfo) => async <T extends APIBody>(path: string, body: Partial<T>): Promise<APIResponseData<T>|APIError<ServiceError>> => {
    const url = `${baseURL}${path}`;
    return fetchApi<Partial<T>, T, ServiceError>(url, {...config, method: 'POST', headers: {...config.headers, 'content-type': 'application/json'}, body});
};
export type Api = {
    getListRequest<T>(path: string, page?: number, size?: number, filter?: string): Promise<APIResponseData<ServiceList<T>>|APIError<ServiceError>>;
    deleteRequest(path: string, ids: string|string[]): Promise<(APIResponseData<ServiceDeleted[]>|APIError<ServiceError>)>;
    createRequest<T extends APIBody>(path: string, body: Partial<T>): Promise<APIResponseData<T>|APIError<ServiceError>>;
};
export const buildApiService = (server: string): Api => {
    const config: APIRequestInfo = {
        method: 'GET',
        headers: {}
    };
    return {
        getListRequest: getListRequest(server, config),
        deleteRequest: deleteRequest(server, config),
        createRequest: createRequest(server, config)
    };
};
