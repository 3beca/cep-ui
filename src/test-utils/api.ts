import {
    EVENT_TYPES_URL,
    TARGETS_URL,
    RULES_URL,
    EVENTS_URL
} from '../services/config';
import {
    ServiceList,
    ServiceError,
    TargetList,
    TargetError,
    EventTypeList,
    EventTypeError,
    Entity,
    Rule,
    RuleTypes,
    RuleList,
    RuleError,
    RuleFilter,
    EventType,
    Target,
    EventLog,
    EventLogList,
    EventLogError
} from '../services/api';
import nock from 'nock';

export const setupNock = (url: string) => {
    const server = nock(url).defaultReplyHeaders({ 'access-control-allow-origin': '*' });
    // Skip Preflight CORS OPTION request
    nock(url).intercept(/./, 'OPTIONS').reply(200, undefined, { 'access-control-allow-origin': '*', 'access-control-allow-headers': 'authorization' }).persist();
    return server;
}

export const generateEntity = function generateListWith(key: string): Entity {
    return {
        id: 'test_entity_' + key
    };
};

export const generateListWith = function generateListWith(many: number = 5, next = false, prev = false, key: string = '_' + many): ServiceList<Entity> {
    const list: ServiceList<Entity> = {
        results: Array.from({length: many},
            (_, idx) => ({
                id: idx + '_' + key
            })
        )
    };
    if (prev) list.prev = 'http://cep/?page=prev';
    if (next) list.next = 'http://cep/?page=next';
    return list;
};

export const generateEventType = (idx: number, key: string, namePrefix: string) => {
    return ({
        id: idx + '_' + key,
        name: namePrefix + 'EventType' + idx,
        url: 'http://cep/elemento' + idx,
        createdAt: '2020-01-01T10:10:10.123Z',
        updatedAt: '2020-01-01T10:10:10.123Z'
    });
};
export const generateEventTypeListWith = (many: number = 5, next = false, prev = false, key: string = '_' + many, namePrefix: string = 'Elemento '): EventTypeList => {
    const list: EventTypeList = {
        results: Array.from({length: many},
            (_, idx) => generateEventType(idx, key, namePrefix)
        )
    };
    if (prev) list.prev = 'http://cep/?page=prev';
    if (next) list.next = 'http://cep/?page=next';
    return list;
};
export const generateTarget = (idx: number, key: string, namePrefix: string) => {
    return (
        {
            id: idx + '_' + key,
            name: namePrefix + 'Target' + idx,
            url: 'http://target/elemento' + idx,
            createdAt: '2020-01-01T10:10:10.123Z',
            updatedAt: '2020-01-01T10:10:10.123Z'
        }
    );
};
export const generateTargetListWith = (many: number = 5, next = false, prev = false, key: string = '_' + many, namePrefix: string = 'Elemento '): TargetList => {
    const list: TargetList = {
        results: Array.from({length: many},
            (_, idx) => generateTarget(idx, key, namePrefix)
        )
    };
    if (prev) list.prev = 'http://cep/?page=prev';
    if (next) list.next = 'http://cep/?page=next';
    return list;
};

const randomType = (index: number): RuleTypes => {
    const types: RuleTypes[] = ['sliding', 'hopping', 'tumbling', 'realtime'];
    const rand = index % 4;
    return types[rand];
};
export const generateRule = (key: string, idx: number, filters: RuleFilter = {temperature: {_eq: 10}, humidity: '45'}): Rule => ({
    id: idx + '_' + key,
    name: 'Rule ' + idx,
    type: randomType(idx),
    eventTypeId: 'eventtypeid',
    eventTypeName: 'EventType ' + idx + '-' + key,
    targetId: 'targetid',
    targetName: 'Target ' + idx + '-' + key,
    skipOnConsecutivesMatches: idx % 2 === 0 ? true : false,
    filters,
    createdAt: '2020-01-01T10:10:10.123Z',
    updatedAt: '2020-01-01T10:10:10.123Z'
});
export const generateRuleListWith = function generateListWith(many: number = 5, next = false, prev = false, key: string = '_' + many): ServiceList<Rule> {
    const list: ServiceList<Rule> = {
        results: Array.from({length: many}, (_, idx) => generateRule(key, idx))
    };
    if (prev) list.prev = 'http://cep/?page=prev';
    if (next) list.next = 'http://cep/?page=next';
    return list;
};

const samplePayload = {
    dateTime: '2020-05-27T18:04:34.182Z',
    device: 'UCAM-01-G-AI00',
    id: '15267:UCAM-01-G-AI00:HUMIDITY',
    type: 'humidity',
    value: 36.58125
};
export const generateEventLog = (key: string, idx: number, payload: any = samplePayload): EventLog => ({
    id: idx + '_' + key,
    eventTypeId: 'eventtypeid',
    eventTypeName: 'EventType ' + idx + '-' + key,
    payload,
    requestId: "53960",
    createdAt: '2020-01-01T10:10:10.123Z'
});
export const generateEventLogListWith = function generateListWith(many: number = 5, next = false, prev = false, key: string = '_' + many): ServiceList<EventLog> {
    const list: ServiceList<EventLog> = {
        results: Array.from({length: many}, (_, idx) => generateEventLog(key, idx))
    };
    if (prev) list.prev = 'http://cep/?page=prev';
    if (next) list.next = 'http://cep/?page=next';
    return list;
};

export const generateEventLogListWithPayload = function generateEventLogListWithPayload(payload: any): ServiceList<EventLog> {
    const list: ServiceList<EventLog> = {
        results: [generateEventLog('EventLogWithPayload', 1, payload)]
    };
    return list;
};

export const serverGet = function serverGetList<T>(server: nock.Scope, path: string, status: number = 200, response: T|ServiceError) {
    return server.get(path).reply(status, response);
};
export const serverGetList = function serverGetList<T>(server: nock.Scope, path: string, page: number = 1, size: number = 10, filter: string = '', status: number = 200, response: ServiceList<T>|ServiceError) {
    return server.get(path + `/?page=${page}&pageSize=${size}${filter ? `&search=${filter}` : ''}`).reply(status, response);
};
export const serverDelete = (server: nock.Scope, path: string, id: string, status: number = 204, response: undefined|ServiceError = undefined) => {
    return server.delete(path + `/${id}`).reply(status, response);
};
export const serverCreate = <T>(server: nock.Scope, path: string, body: string, status: number = 204, response: T|ServiceError) => {
    return server.post(path, body).reply(status, response);
};

export const serverGetEventTypeList = (server: nock.Scope, page: number = 1, size: number = 10, filter= '', status: number = 200, response: EventTypeList|EventTypeError = generateEventTypeListWith(10, false, false)) => {
    return serverGetList(server, EVENT_TYPES_URL, page, size, filter, status, response);
};
export const serverDeleteEventType = (server: nock.Scope, eventTypeId: string, status: number = 204, response: undefined|EventTypeError = undefined) => {
    return serverDelete(server, EVENT_TYPES_URL, eventTypeId, status, response);
};
export const serverCreateEventType = (server: nock.Scope, body: Partial<EventType>, status: number = 200, response: EventType|ServiceError) => {
    return serverCreate(server, EVENT_TYPES_URL, JSON.stringify(body), status, response);
};

export const serverGetTargetList = (server: nock.Scope, page: number = 1, size: number = 10, filter= '', status: number = 200, response: TargetList|TargetError = generateTargetListWith(10, false, false)) => {
    return serverGetList(server, TARGETS_URL, page, size, filter, status, response);
};
export const serverDeleteTarget = (server: nock.Scope, targetId: string, status: number = 204, response: undefined|TargetError = undefined) => {
    return serverDelete(server, TARGETS_URL, targetId, status, response);
};
export const serverCreateTarget = (server: nock.Scope, body: Partial<Target>, status: number = 200, response: Target|ServiceError) => {
    return serverCreate(server, TARGETS_URL, JSON.stringify(body), status, response);
};

export const serverGetRuleList = (server: nock.Scope, page: number = 1, size: number = 10, filter= '', status: number = 200, response: RuleList|RuleError = generateRuleListWith(10, false, false)) => {
    return serverGetList(server, RULES_URL, page, size, filter, status, response);
};
export const serverDeleteRule = (server: nock.Scope, ruleId: string, status: number = 204, response: undefined|RuleError = undefined) => {
    return serverDelete(server, RULES_URL, ruleId, status, response);
};
export const serverCreateRule = (server: nock.Scope, body: Partial<Rule>, status: number = 200, response: Rule|ServiceError) => {
    return serverCreate(server, RULES_URL, JSON.stringify(body), status, response);
};

export const serverGetEventLogList = (server: nock.Scope, page: number = 1, size: number = 10, eventTypeId= '', status: number = 200, response: EventLogList|EventLogError = generateEventLogListWith(10, false, false)) => {
    return server.get(EVENTS_URL + `/?page=${page}&pageSize=${size}${eventTypeId ? `&eventTypeId=${eventTypeId}` : ''}`).reply(status, response);
};
export const serverGet401 = (server: nock.Scope, path: string) => {
    return server.get(path).reply(401, {error: 'missing authorization header'});
};
export const serverGetList401 = (server: nock.Scope, path: string, page = 1, size = 10, filter?: string) => {
    return server.get(path + `/?page=${page}&pageSize=${size}${filter ? `&search=${filter}` : ''}`).reply(401, {error: 'missing authorization header'});
};
export const serverPost401 = (server: nock.Scope, path: string, body: string) => {
    return server.post(path, body).reply(401, {error: 'missing authorization header'});
};
export const serverDelete401 = (server: nock.Scope, path: string, id: string) => {
    return server.delete(path + `/${id}`).reply(401, {error: 'missing authorization header'});
};

export const serverGetAuth = function serverGetList<T>(server: nock.Scope, path: string, key: string,status: number = 200, response: T|ServiceError) {
    return server.get(
        path,
        undefined,
        {reqheaders: {Authorization: 'apiKey ' + key}}
    ).reply(status, response);
};

export const serverGetListAuth = function serverGetList<T>(server: nock.Scope, path: string, key: string, page: number = 1, size: number = 10, filter: string = '', status: number = 200, response: ServiceList<T>|ServiceError) {
    return server.get(
        path + `/?page=${page}&pageSize=${size}${filter ? `&search=${filter}` : ''}`,
        undefined,
        {reqheaders: {Authorization: 'apiKey ' + key}}
    ).reply(status, response);
};

export const serverDeleteAuth = (server: nock.Scope, path: string, key: string, id: string, status: number = 204, response: undefined|ServiceError = undefined) => {
    return server.delete(
        path + `/${id}`,
        undefined,
        {reqheaders: {Authorization: 'apiKey ' + key}}
    ).reply(status, response);
};
export const serverCreateAuth = <T>(server: nock.Scope, path: string, key: string, body: string, status: number = 204, response: T|ServiceError) => {
    return server.post(
        path,
        body,
        {reqheaders: {Authorization: 'apiKey ' + key}}
    ).reply(status, response);
};
