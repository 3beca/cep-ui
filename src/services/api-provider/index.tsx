import * as React from 'react';
import { APIRequestInfo, isAPIError } from '../../utils/fetch-api';
import { buildApiService } from '../api';
import { VersionInfo } from '../api/models';
import { BASE_URL, VERSION_URL } from '../config';
import { loadApikey, saveApikey, clearApikey } from '../../utils';
import {
    APIContextState,
    ApiContextActions,
    ValidationState,
    initialContext,
    APIContext,
    UpdateAPIContext
} from './api-context';

// config = {method: 'GET', headers: {'authorization': 'apiKey ' + apiKey}};
export const apiReducer = (
    state: APIContextState,
    action: ApiContextActions
) => {
    switch (action.type) {
        case 'VALIDATING': {
            return {
                isValidating: true,
                isValidated: ValidationState.PENDING,
                isValid: false
            };
        }
        case 'SERVER_NOT_FOUND': {
            return {
                isValidating: false,
                isValidated: ValidationState.NOT_FOUND,
                isValid: false,
                invalidReason: `${action.url} not found`
            };
        }
        case 'NO_REQUIRE_APIKEY': {
            return {
                isValidating: false,
                isValidated: ValidationState.VALIDATED,
                isValid: true,
                requireKey: false,
                api: buildApiService(BASE_URL),
                version: action.version
            };
        }
        case 'REQUIRE_APIKEY': {
            return {
                isValidating: false,
                isValidated: ValidationState.VALIDATED,
                isValid: false,
                requireKey: true,
                invalidReason: action.reason
            };
        }
        case 'SET_APIKEY': {
            const apiKey = action.apiKey;
            const version = action.version;
            const apiInfo: APIRequestInfo = {
                headers: { Authorization: 'apiKey ' + apiKey },
                method: 'GET'
            };
            return {
                isValidating: false,
                isValidated: ValidationState.VALIDATED,
                isValid: true,
                requireKey: true,
                apiKey,
                api: buildApiService(BASE_URL, apiInfo),
                version
            };
        }
        default:
            return state;
    }
};
export const APIProvider: React.FC<{}> = props => {
    const [state, dispatch] = React.useReducer(apiReducer, initialContext);

    const checkApikey = React.useCallback(async (apiKey: string) => {
        dispatch({ type: 'VALIDATING' });
        const apiInfo: APIRequestInfo = {
            headers: { Authorization: 'apiKey ' + apiKey },
            method: 'GET'
        };
        const api = buildApiService(BASE_URL, apiInfo);
        const responseWithtoken = await api.getRequest<VersionInfo>(
            VERSION_URL
        );
        if (isAPIError(responseWithtoken)) {
            clearApikey();
            dispatch({
                type: 'REQUIRE_APIKEY',
                reason: `ApiKey ${apiKey} is NOT valid`
            });
            return;
        } else {
            saveApikey(apiKey);
            dispatch({
                type: 'SET_APIKEY',
                apiKey,
                version: responseWithtoken.data.version
            });
            return;
        }
    }, []);

    const fireValidation = React.useCallback(async () => {
        dispatch({ type: 'VALIDATING' });
        let api = buildApiService(BASE_URL);
        const response = await api.getRequest<VersionInfo>(VERSION_URL);
        if (isAPIError(response)) {
            if (response.errorCode === 500) {
                dispatch({ type: 'SERVER_NOT_FOUND', url: BASE_URL });
                return;
            }
            const apiKey = loadApikey();
            if (!apiKey) {
                dispatch({
                    type: 'REQUIRE_APIKEY',
                    reason: 'apiKey not found'
                });
                return;
            }
            checkApikey(apiKey);
        } else {
            dispatch({
                type: 'NO_REQUIRE_APIKEY',
                version: response.data.version
            });
        }
    }, [checkApikey]);

    const invalidateApiKey = React.useCallback(() => {
        clearApikey();
        fireValidation();
    }, [fireValidation]);

    const setApiKey = React.useCallback(
        (apiKey: string) => {
            checkApikey(apiKey);
        },
        [checkApikey]
    );

    const utils = React.useMemo(
        () => ({
            invalidateApiKey,
            setApiKey
        }),
        [invalidateApiKey, setApiKey]
    );

    React.useEffect(() => {
        fireValidation();
    }, [fireValidation]);

    return (
        <APIContext.Provider value={state}>
            <UpdateAPIContext.Provider value={utils} {...props} />
        </APIContext.Provider>
    );
};

export const useUpdateAPIProvider = () => {
    return React.useContext(UpdateAPIContext);
};

export const useAPIProviderStatus = () => {
    const api = React.useContext(APIContext);
    return {
        showLoading:
            api.isValidating || api.isValidated === ValidationState.PENDING,
        showNoService:
            !api.isValidating && api.isValidated === ValidationState.NOT_FOUND,
        showLogin:
            !api.isValidating &&
            api.requireKey &&
            api.isValidated === ValidationState.VALIDATED &&
            !api.isValid,
        requireApikey: api.requireKey,
        invalidReason: api.invalidReason,
        apiKey: api.apiKey,
        version: api.version
    };
};

export const useAPIProvider = () => {
    const api = React.useContext(APIContext);
    if (!api.api) throw Error('no API available');
    return {
        api: api.api,
        apiKey: api.apiKey,
        version: api.version
    };
};
