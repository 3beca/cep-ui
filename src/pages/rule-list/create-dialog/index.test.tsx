import * as React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import CreateRuleDialog from './index';
import { RuleTypes } from '../../../services/api';

test('CreateRuleDialog should render null, no icon button, no dialog', () => {
    // Render null, no icon button, no dialog
    const {queryByLabelText} = render(<CreateRuleDialog isOpen={false}/>);
    expect(queryByLabelText(/create rule dialog/i)).not.toBeInTheDocument();
});

test('CreateRuleDialog should render the create rules dialog', async () => {
    const onClose = jest.fn();

    // Render dialog
    const {container, getByLabelText, getByText} = render(
        <CreateRuleDialog isOpen={true}/>
    );
    getByLabelText(/create rule dialog/i);
    expect(onClose).toBeCalledTimes(0);
    getByText(/close/i);
    expect(getByText(/^select$/i)).toBeDisabled();
    expect(container).toMatchSnapshot();

});

test('CreateRuleDialog should render dialog when click in icon dialog and close when press close button', async () => {
    const onClose = jest.fn();

    // Render dialog
    const {getByLabelText, queryByLabelText, getByText, rerender} = render(
        <CreateRuleDialog isOpen={true} onClose={onClose}/>
    );
    getByLabelText(/create rule dialog/i);
    expect(onClose).toBeCalledTimes(0);
    const closeButton = getByText(/close/i);

    // close Dialog
    fireEvent.click(closeButton);
    expect(onClose).toBeCalledTimes(1);

    rerender(<CreateRuleDialog isOpen={false} onClose={onClose}/>);
    await waitFor(() => expect(queryByLabelText(/create rule dialog/i)).not.toBeInTheDocument());
});

const runSelectCardTest = (ariaLabel: RegExp, typeSelected: RuleTypes) => {
    test('CreateRuleDialog should render dialog when click in icon dialog and close when select ' + ariaLabel, async () => {
        const onClose = jest.fn();
        const onSelected = jest.fn();

        // Render dialog
        const {getByLabelText, queryByLabelText, getByText, rerender} = render(
            <CreateRuleDialog isOpen={true} onClose={onClose} onSelect={onSelected}/>
        );
        getByLabelText(/create rule dialog/i);
        expect(onClose).toBeCalledTimes(0);
        getByText(/close/i);
        const selectButton = getByText(/^select$/i);
        const cardRule = getByLabelText(ariaLabel);
        expect(selectButton).toBeDisabled();

        // Select rule
        fireEvent.click(cardRule);
        expect(selectButton).not.toBeDisabled();

        // Confirm select
        fireEvent.click(selectButton);
        expect(onClose).toBeCalledTimes(1);
        expect(onSelected).toHaveBeenNthCalledWith(1, typeSelected);

        rerender(<CreateRuleDialog isOpen={false} onClose={onClose}/>);
        await waitFor(() => expect(queryByLabelText(/create rule dialog/i)).not.toBeInTheDocument());
    });
};

runSelectCardTest(/create rule real time card/i, 'none');
runSelectCardTest(/create rule hopping card/i, 'hopping');
runSelectCardTest(/create rule sliding card/i, 'sliding');
runSelectCardTest(/create rule tumbling card/i, 'tumbling');