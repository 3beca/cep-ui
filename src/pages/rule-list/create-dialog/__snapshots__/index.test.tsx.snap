// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateRuleDialog should render the create rules dialog 1`] = `
<div
  aria-hidden="true"
/>
`;
